### PR TITLE
Syncronize access when updating config on search layer

### PR DESCRIPTION
**Pull request by nronas**
#### Summary
Fixes a race condition due to a non synchronized access of a critical section, in our case updating the config of the search layer. 

Solution: synchronise access via making config to be an `atomic.Value`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-30868

#### Release Note

```release-note
Fix race condition when updating config on search layer.
```

### CODE REVIEW

File: store/searchlayer/layer.go [hunk 1]
```diff
@@ -18,6 +20,7 @@ type SearchStore struct {
 	channel      *SearchChannelStore
 	post         *SearchPostStore
 	config       *model.Config
+	configLock   sync.RWMutex
```
Comment (nronas) [inline store/searchlayer/layer.go:23]: Also a `sync.Mutex` here will be just fine.

Comment (streamer45) [reply store/searchlayer/layer.go:23]: I am wondering if using `atomic.Value` would be more appropriate in this case. We use that consistently in the `app.Server` struct.

File: store/searchlayer/layer_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// See LICENSE.txt for license information.
+
+package searchlayer_test
+
+import (
+	"os"
+	"testing"
+
+	"github.com/mattermost/mattermost-server/v5/store/searchlayer"
+
```
Comment (agnivade) [inline store/searchlayer/layer_test.go:11]: Nit: there is no need for gaps here and on L16

File: store/searchlayer/layer_test.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	cfg := &model.Config{}
+	cfg.SetDefaults()
+	cfg.ClusterSettings.MaxIdleConns = model.NewInt(1)
+	searchEngine := searchengine.NewBroker(cfg, nil)
+	layer := searchlayer.NewSearchLayer(&testlib.TestStore{Store: store}, searchEngine, cfg)
+
+	for i := 0; i < 5; i++ {
+		go func() {
+			layer.UpdateConfig(cfg.Clone())
+		}()
```
Comment (agnivade) [inline store/searchlayer/layer_test.go:39]: I think it might be better to wait for the goroutines to finish before exiting the test using a waitgroup.

**Review by agnivade — COMMENTED** [review]
Looks good. Just a few small comments.

File: store/searchlayer/layer_test.go (hunk 2 above)
Comment (nronas) [reply store/searchlayer/layer_test.go:11]: damn this is my editor doing `goimports` on save.

File: store/searchlayer/layer_test.go (hunk 3 above)
Comment (nronas) [reply store/searchlayer/layer_test.go:39]: nice catch, totally missed that

**Review by agnivade — APPROVED** [review]
LGTM

**Review by streamer45 — APPROVED** [review]

**Pull request merged**